# Q2: Top Products and Customer Concentration Analysis

### Business Question
1. Which products contribute most to revenue?
2. Is revenue concentrated among a small group of customers?

### Objective
Evaluate product-level revenue distribution and assess customer concentration risk.

In [51]:
import pandas as pd
df = pd.read_csv('../data/processed/retail_cleaned.csv')
df.head()

C:\Users\kitki\AppData\Local\Temp\ipykernel_2168\1877642722.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/processed/retail_cleaned.csv')


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer_ID,Country,TotalSales
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [52]:
## Determine the top 10 products by total sales
total_sales_by_stockcode = df.groupby('StockCode')['TotalSales'].sum().sort_values(ascending=False).head(10)
print(total_sales_by_stockcode)

StockCode
22423     344563.25
M         340716.28
DOT       322657.48
85123A    263109.67
85099B    183454.83
23843     168469.60
47566     149187.05
84879     132187.92
POST      127597.42
22086     123141.54
Name: TotalSales, dtype: float64


In [53]:
total_revenue = df['TotalSales'].sum()
print(total_revenue)

20972594.568000007


In [54]:
top_products_share = (total_sales_by_stockcode / total_revenue) * 100
print(top_products_share)

StockCode
22423     1.642921
M         1.624579
DOT       1.538472
85123A    1.254540
85099B    0.874736
23843     0.803284
47566     0.711343
84879     0.630289
POST      0.608401
22086     0.587155
Name: TotalSales, dtype: float64


In [55]:
top_products_share.sum()

10.275719739932557

In [56]:
## Exclude non-product items and recalculate the top 10 products by total sales
exclude_codes = ['POST', 'M', 'DOT']
df_filtered = df[~df['StockCode'].isin(exclude_codes)]
total_sales_by_stockcode_filtered = df_filtered.groupby('StockCode')['TotalSales'].sum().sort_values(ascending=False).head(10)
print(total_sales_by_stockcode_filtered)

StockCode
22423     344563.25
85123A    263109.67
85099B    183454.83
23843     168469.60
47566     149187.05
84879     132187.92
22086     123141.54
79321      85489.91
23166      81700.92
22197      80920.64
Name: TotalSales, dtype: float64


In [57]:
total_filtered_revenue = df_filtered['TotalSales'].sum()
print(total_filtered_revenue)

20181623.387999997


In [58]:
top_filtered_products_share = (total_sales_by_stockcode_filtered / total_filtered_revenue) * 100
print(top_filtered_products_share)

StockCode
22423     1.707312
85123A    1.303709
85099B    0.909019
23843     0.834767
47566     0.739222
84879     0.654992
22086     0.610167
79321     0.423603
23166     0.404828
22197     0.400962
Name: TotalSales, dtype: float64


In [59]:
top_filtered_products_share.sum()

7.988580992739315

## Top Products by Revenue

### Method
Group by StockCode and calculate total revenue. Exclude non-product codes such as POST, DOT, and M.

### Insight
The top 10 products account for approximately 8% of total product revenue, indicating a diversified product portfolio.


In [60]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer_ID,Country,TotalSales
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [61]:
## Analyze top customers by total sales
df_customers = df[df['Customer_ID'].notna()]

In [62]:
top_customers = df_customers.groupby('Customer_ID')['TotalSales'].sum().sort_values(ascending=False).head(10)
total_customer_revenue = df_customers['TotalSales'].sum()
top_customers_share = (top_customers/ total_customer_revenue) * 100
print(top_customers_share)


Customer_ID
18102.0    3.431251
14646.0    2.979145
14156.0    1.769367
14911.0    1.668069
17450.0    1.391913
13694.0    1.107355
17511.0    0.989682
16446.0    0.949492
16684.0    0.829280
12415.0    0.814151
Name: TotalSales, dtype: float64


In [63]:
top_customers_share.sum()

15.929706888364821

## Customer Revenue Concentration

### Method
Group by Customer_ID and calculate total revenue per customer. Compute revenue share for top customers.

### Insight
The top 10 customers contribute approximately 16% of total identified customer revenue, suggesting limited dependency risk.


In [64]:
## Pareto analysis of customer revenue concentration
customer_revenue = df_customers.groupby('Customer_ID')['TotalSales'].sum()
customer_revenue_sorted = customer_revenue.sort_values(ascending=False)
cumulative_revenue = customer_revenue_sorted.cumsum()
cumulative_percentage = (cumulative_revenue / total_customer_revenue) * 100
print(cumulative_percentage.head(20))

Customer_ID
18102.0     3.431251
14646.0     6.410397
14156.0     8.179763
14911.0     9.847832
17450.0    11.239745
13694.0    12.347101
17511.0    13.336783
16446.0    14.286275
16684.0    15.115556
12415.0    15.929707
15061.0    16.706437
16029.0    17.395194
17949.0    18.063769
15311.0    18.721878
13089.0    19.379800
12931.0    19.900259
14298.0    20.414221
15769.0    20.914149
12346.0    21.351248
13798.0    21.778657
Name: TotalSales, dtype: float64


In [65]:
total_customers = len(customer_revenue_sorted)
print(f"Total customers: {total_customers}")

Total customers: 5878


In [66]:
top_20_percent_customers = int(total_customers * 0.2)

In [50]:
cumulative_percentage.iloc[top_20_percent_customers]

77.25916970997358

## Pareto Analysis Result
Pareto analysis reveals that the top 20% of customers contribute approximately 77% of total revenue. This indicates a structurally concentrated customer base consistent with a typical 80/20 distribution.


## Preliminary Findings

(Product diversification)
(Customer concentration – Pareto ~77%)